File: SOC/src/cc/game/soc/core/AITuning.java
package cc.game.soc.core;

import java.util.Map;

public abstract class AITuning {

	private static AITuning instance = null;
	
	public static AITuning getInstance() {
		if (instance == null) {
			instance = new AITuning() {
				public double getScalingFactor(String p) {
					return 1.0;
				}
			};
		}
		return instance;
	}
	
	public static void setInstance(AITuning instance) {
		AITuning.instance = instance;
	}

	public static void setMap(final Map<String, Double> map) {
	    instance = new AITuning() {
            @Override
            public double getScalingFactor(String property) {
                return map.get(property);
            }
        };
    }

	public abstract double getScalingFactor(String property);
	
}

File: SOC/src/cc/game/soc/core/AITuning.kt
package cc.game.soc.core

abstract class AITuning {
	abstract fun getScalingFactor(property: String): Double

	companion object {
		private var instance: AITuning? = null
		fun getInstance(): AITuning {
			return instance?:object : AITuning() {
					init {
						instance = this
					}
					override fun getScalingFactor(p: String): Double {
						return 1.0
					}
				}
			}


		fun setInstance(instance: AITuning) {
			Companion.instance = instance
		}

		fun setMap(map: Map<String, Double>) {
			instance = object : AITuning() {
				override fun getScalingFactor(property: String): Double {
					return map[property]?:1.0
				}
			}
		}
	}
}